pkg/auth/auth.go

```diff
@@ -0,0 +1,250 @@
+			c.Cloud.ActiveDirectoryAuthorityHost, c.AuthConfig.TenantID)
+		if err != nil {
+			return nil, err
+		}
+		token, err = adal.NewServicePrincipalToken(
+			*oauthConfig, c.AuthConfig.ClientID, c.AuthConfig.ClientSecret, resource)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	auth = autorest.NewBearerAuthorizer(token)
+	return auth, err
+}
+
+// SetAuthConfig sets the configuration needed for Authentication
+func (c *Config) SetAuthConfig() error {
+	var err error
+	c.Cloud = cloud.AzurePublic
+
+	if authFilepath := os.Getenv("AZURE_AUTH_LOCATION"); authFilepath != "" {
+		auth := &Authentication{}
+		err = auth.NewAuthenticationFromFile(authFilepath)
+		if err != nil {
+			return err
+		}
+		c.AuthConfig = auth
+	}
+
+	if aksCredFilepath := os.Getenv("AKS_CREDENTIAL_LOCATION"); aksCredFilepath != "" {
+		c.AKSCredential, err = NewAKSCredential(aksCredFilepath)
+		if err != nil {
+			return err
+		}
+
+		if c.AKSCredential != nil {
+			var clientId string
+			if !strings.EqualFold(c.AKSCredential.ClientID, "msi") {
+				clientId = c.AKSCredential.ClientID
+			}
+
+			//Set Azure cloud environment
+			c.Cloud = getCloudConfiguration(c.AKSCredential.Cloud)
+			c.AuthConfig = NewAuthentication(
+				clientId,
+				c.AKSCredential.ClientSecret,
+				c.AKSCredential.SubscriptionID,
+				c.AKSCredential.TenantID,
+				c.AKSCredential.UserAssignedIdentityID)
+		}
+
+		if clientID := os.Getenv("AZURE_CLIENT_ID"); clientID != "" {
+			c.AuthConfig.ClientID = clientID
+		}
+
+		if clientSecret := os.Getenv("AZURE_CLIENT_SECRET"); clientSecret != "" {
+			c.AuthConfig.ClientSecret = clientSecret
+		}
+
+		if userIdentityClientId := os.Getenv("VIRTUALNODE_USER_IDENTITY_CLIENTID"); userIdentityClientId != "" {
+			c.AuthConfig.UserIdentityClientId = userIdentityClientId
+		}
+		c.AuthConfig.UseUserIdentity = (len(c.AuthConfig.ClientID) == 0)
+
+		if c.AuthConfig.UseUserIdentity {
+			if len(c.AuthConfig.UserIdentityClientId) == 0 {
+				return fmt.Errorf("neither AZURE_CLIENT_ID or VIRTUALNODE_USER_IDENTITY_CLIENTID is being set")
+			}
+
+			log.G(context.TODO()).Info("Using user identity for Authentication")
+		}
+
+		if tenantID := os.Getenv("AZURE_TENANT_ID"); tenantID != "" {
+			c.AuthConfig.TenantID = tenantID
+		}
+
+		if subscriptionID := os.Getenv("AZURE_SUBSCRIPTION_ID"); subscriptionID != "" {
+			c.AuthConfig.SubscriptionID = subscriptionID
+		}
+	}
+
+	resource := c.Cloud.Services[cloud.ResourceManager].Endpoint
+
+	c.Authorizer, err = c.getAuthorizer(resource)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Authentication represents the Authentication file for Azure.
+type Authentication struct {
+	ClientID             string `json:"clientId,omitempty"`
+	ClientSecret         string `json:"clientSecret,omitempty"`
+	SubscriptionID       string `json:"subscriptionId,omitempty"`
+	TenantID             string `json:"tenantId,omitempty"`
+	UseUserIdentity      bool   `json:"useUserIdentity,omitempty"`
```

Comment:
UseUserIdentity flag can be derived from the length of the clientID field. You may not need this field in the struct.

Reply:
This is how the code determines the value [here](https://github.com/virtual-kubelet/azure-aci/pull/299/files/7563cac2f7a24ceafb6f8a6859dc0e610e01bb1a#diff-0ccbf9a2a7d709b87dbf2f30cc730528c6299d537bf721300d49502cfe521c0aR113).

Reply:
I mean you don't need a field. You can expose a method
func (a Authentication) UseUserIdentity bool {
    return len(c.AuthConfig.ClientID) == 0
}
whenever you want to check this boolean.

Reply:
done.